Mon Aug 18 16:23:16 CEST 2003  Jan Hubicka  <jh@suse.cz>

	PR c/11639
	* i386.c (ix86_expand_carry_flag_compare): Validate operand.

--- gcc-3.3.1/gcc/config/i386/i386.c.pr11639	2003-09-01 11:24:00.000000000 +0200
+++ gcc-3.3.1/gcc/config/i386/i386.c	2003-09-01 11:24:55.000000000 +0200
@@ -9587,6 +9587,14 @@ ix86_expand_carry_flag_compare (code, op
     default:
       return false;
     }
+
+  /* Swapping operands may cause constant to appear as first operand.  */
+  if (!nonimmediate_operand (op0, VOIDmode))
+    {
+      if (no_new_pseudos)
+	return false;
+      op0 = force_reg (mode, op0);
+    }
   ix86_compare_op0 = op0;
   ix86_compare_op1 = op1;
   *pop = ix86_expand_compare (code, NULL, NULL);
